plat/common/include/uk/plat/common/memory.h

```diff
@@ -112,6 +117,80 @@ ukplat_memregion_list_insert(struct ukplat_memregion_list *list,
 	return (int)i;
 }
 
+/**
+ * Insert a new region into memory region list. This extends
+ * ukplat_memregion_list_insert to carve out the area of any pre-existing
+ * overlapping regions. The virtual addresses will leave out any carved out
+ * region.
+ * If there is not enough space for all resulting region, the function will
+ * insert as many as possible and then signal return an error.
+ * @param list
+ *   The memory region list to insert the range into
+ * @param mrd
+ *   The memory range to insert
+ * @param min_size
+ *   The minimum size a inserted region has to have. Regions smaller than this
+ *   size won't be added. Settings this parameter zero will disable this
+ *   behavior.
+ * @returns
+ *   zero if the operation was successful, an negative errno otherwise
+ */
+static inline int
+ukplat_memregion_list_insert_split_phys(struct ukplat_memregion_list *list,
+					const struct ukplat_memregion_desc *mrd,
+					__sz min_size)
+{
+	struct ukplat_memregion_desc *mrdp;
+	struct ukplat_memregion_desc mrdc;
+	__paddr_t pstart, pend;
+	__vaddr_t voffset;
+	int i;
+	int rc;
+
+	voffset = mrd->vbase - mrd->pbase;
+	pstart = mrd->pbase;
+	pend = mrd->pbase + mrd->len;
+
+	memcpy(&mrdc, mrd, sizeof(mrdc));
+
+	for (i = 0; i < (int)list->count; i++) {
+		ukplat_memregion_get(i, &mrdp);
+		if (!ukplat_memregion_desc_overlap(mrdp, pstart, pend))
+			continue;
```

Comment:
Since the list is sorted, I believe we can break the iteration just when we find the first element that has: 
`mrdp->pbase > pend`